src/types.rs

```diff
@@ -764,3 +776,178 @@ impl LateLintPass for AbsurdExtremeComparisons {
         }
     }
 }
+
+/// **What it does:** This lint checks for comparisons where the relation is always either true or false, but where one side has been upcast so that the comparison is necessary. Only integer types are checked.
+///
+/// **Why is this bad?** An expression like `let x : u8 = ...; (x as u32) > 300` will mistakenly imply that it is possible for `x` to be outside the range of `u8`.
+///
+/// **Known problems:** None
+///
+/// **Example:** `let x : u8 = ...; (x as u32) > 300`
+declare_lint! {
+    pub INVALID_UPCAST_COMPARISONS, Warn,
+    "a comparison involving an upcast which is always true or false"
+}
+
+pub struct InvalidUpcastComparisons;
+
+impl LintPass for InvalidUpcastComparisons {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(INVALID_UPCAST_COMPARISONS)
+    }
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
```

Comment:
Ah, sorry- misunderstood you when you said I could derive that. You were just referring to `Eq`, not  also `PartialEq`, right?